File: solved/word-search.ts
/**
 * https://leetcode.com/problems/word-search/
 */
export { exist };

function exist(board: string[][], word: string): boolean {
  const lastXIndex = board.length - 1;
  const lastYIndex = board[0].length - 1;

  function dfs(x: number, y: number, remaining: string[], visited: Set<string>): boolean {
    if (0 > x || x > lastXIndex) return false;
    if (0 > y || y > lastYIndex) return false;
    if (remaining[0] !== board[x][y]) return false;
    if (remaining[0] === board[x][y] && remaining.length === 1) return true;

    const nextRemaining = remaining.slice(1);

    const rightKey = `${x + 1},${y}`;
    if (!visited.has(rightKey)) {
      visited.add(rightKey);
      if (dfs(x + 1, y, nextRemaining, visited)) return true;
      visited.delete(rightKey);
    }

    const leftKey = `${x - 1},${y}`;
    if (!visited.has(leftKey)) {
      visited.add(leftKey);
      if (dfs(x - 1, y, nextRemaining, visited)) return true;
      visited.delete(leftKey);
    }

    const upKey = `${x},${y + 1}`;
    if (!visited.has(upKey)) {
      visited.add(upKey);
      if (dfs(x, y + 1, nextRemaining, visited)) return true;
      visited.delete(upKey);
    }

    const downKey = `${x},${y - 1}`;
    if (!visited.has(downKey)) {
      visited.add(downKey);
      if (dfs(x, y - 1, nextRemaining, visited)) return true;
      visited.delete(downKey);
    }

    return false;
  }

  const boardSize = board.length * board[0].length;
  if (word.length > boardSize) return false;

  const characters = word.split('');

  for (let i = 0; i <= lastXIndex; i++) {
    for (let j = 0; j <= lastYIndex; j++) {
      if (dfs(i, j, characters, new Set())) return true;
    }
  }

  return false;
}

File: python/today5/word_search.py
# https://leetcode.com/problems/word-search/description/
from typing import List

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        row_count = len(board)
        col_count = len(board[0])
        visited = set()
        def dfs(x: int, y: int, k: 0) -> bool:
            if k == len(word):
                return True
            
            if x < 0 or x >= row_count or y < 0 or y >= col_count:
                return False
            
            if (x,y) in visited:
                return False

            if word[k] != board[x][y]:
                return False
            
            visited.add((x,y))
            up = dfs(x, y-1, k+1)
            down = dfs(x, y+1, k+1)
            left = dfs(x-1, y, k+1)
            right = dfs(x+1, y, k+1)
            visited.remove((x,y))

            return up or down or left or right

        for x in range(row_count):
            for y in range(col_count):
                if dfs(x,y,0):
                    return True
        return False
